keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -125,65 +126,111 @@ As an end user using encryption at rest, whenever the key change is detected we
-        Version string `json:"version,omitempty"`
-        // The name of the resource.
-        Resource string `json:"resource,omitempty"`
+      // The name of the group.
+      Group string `json:"group,omitempty" protobuf:"bytes,1,opt,name=group"`
+      // The name of the version.
+      Version string `json:"version,omitempty" protobuf:"bytes,2,opt,name=version"`
+      // The name of the resource.
+      Resource string `json:"resource,omitempty" protobuf:"bytes,3,opt,name=resource"`
     }
-    
+
+    type MigrationConditionType string
+
+    const (
+      // Indicates that the migration is running.
+      MigrationRunning MigrationConditionType = "Running"
+      // Indicates that the migration has completed successfully.
+      MigrationSucceeded MigrationConditionType = "Succeeded"
+      // Indicates that the migration has failed.
+      MigrationFailed MigrationConditionType = "Failed"
+    )
+
+    // Describes the state of a migration at a certain point.
+    type MigrationCondition struct {
+      // Type of the condition.
+      Type MigrationConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=MigrationConditionType"`
+      // Status of the condition, one of True, False, Unknown.
+      Status corev1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
+      // The last time this condition was updated.
+      // +optional
+      LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty" protobuf:"bytes,3,opt,name=lastUpdateTime"`
+      // The reason for the condition's last transition.
+      // +optional
+      Reason string `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+      // A human readable message indicating details about the transition.
+      // +optional
+      Message string `json:"message,omitempty" protobuf:"bytes,5,opt,name=message"`
+    }
+
     // Status of the storage version migration.
     type StorageVersionMigrationStatus struct {
-        // The latest available observations of the migration's current state.
-        // +optional
-        // +patchMergeKey=type
-        // +patchStrategy=merge
-        Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+      // The latest available observations of the migration's current state.
+      // +patchMergeKey=type
+      // +patchStrategy=merge
+      // +listType=map
+      // +listMapKey=type
+      // +optional
+      Conditions []MigrationCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
+      // ResourceVersion to compare with the GC cache for performing the migration.
+      ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,2,opt,name=resourceVersion"`
+      // LastMigratedResourceNameHash is use to pick up migration from where it left off in case of failure.
```

Comment:
+1 to a hash versus exposing directly.  I'll read further down to see how the hash is computed

Reply:
I can't find where this is described. How is it used?

Reply:
This came up in my discussion with @enj. The thought here is that users may have sensitive or secret projects, and those names might be reflected in the resources they will be creating.

Reply:
> I can't find where this is described. How is it used?

I haven't added the hashing details yet. But before beginning the migration, we will check if this field is set. If it is set, it means the controller needs to continue from where it stopped, crashed, or encountered an error last time. If it is not set, it means this will be a completely new migration. While carrying out the migration, I'm sorting keys [lexicographically](https://github.com/nilekhc/kubernetes/commit/90b894f3aed7dbfd1138598a541fcb984002fcad#diff-c2c55b6f042b05145512cca28308b8b5e8ee67f062ee79f818c3ea67954ebb2dR196-R197). So, if the controller needs to continue from where it was left off, we can skip until we find the hash set in the status and then continue from there.

Reply:
> So, if the controller needs to continue from where it was left off, we can skip until we find the hash set in the status and then continue from there.

What if that item was deleted?

Reply:
We don't care if the item is dropped, do we? Deleted item doesn't need migration, so if item is not found while performing the migration, we simply move on to the next.

Reply:
> We don't care if the item is deleted, do we? Deleted item doesn't need migration, so if item is not found while performing the migration, we simply move on to the next.

I'm not clear on how you could look at a hash of namespace/name and determine if item is after a hash you didn't observe.

Reply:
At time t0 when migration is started, we obtain a sorted list of resources to be migrated. If any new resource is added at time t1, we don't need to migrate it as it would already be at the desired version. Therefore, even if the migration starts at t0 and the controller is subsequently restarted, we can simply [skip](https://github.com/nilekhc/kubernetes/commit/3159bedf25bd4f8401b8094806909cbc97dbffa3#diff-c2c55b6f042b05145512cca28308b8b5e8ee67f062ee79f818c3ea67954ebb2dR211-R221) until the hashed resource and then resume migration. After the controller restart, when we receive the sorted list again, any _"new"_ resources added before the hashed one should already be at the desired version and can therefore be skipped.

Reply:
> What if that item was deleted?

Oh, I see what you mean, how about we store index of last process key along with hash? That way if we can't find hashed key, we can start processing from the next indexed from the stored one. The only downside is that we could end up processing some newer or already processed items, but it should be ok since we are making no-op update anyways.

Reply:
Exposing names seems like a bad idea PII-wise from a different API. How about storing the first letter or two of the last seen item every so-often?  That would give us the ability to place ourselves in the list pretty granularly, but two letters isn't sensitive

Reply:
This could work. We might end reprocessing some items, but I think that's ok.